modules/tinymce/src/core/main/ts/html/Sanitization.ts

```diff
@@ -239,21 +239,40 @@ const sanitizeMathmlElement = (node: Element, settings: DomParserSettings) => {
     return hasAllowedEncodings && Type.isString(encoding) && Arr.contains(allowedEncodings, encoding);
   };
 
-  purify.addHook('uponSanitizeElement', (node, evt) => {
-    const lcTagName = evt.tagName ?? node.nodeName.toLowerCase();
-
+  const isValidElementOpt = (lcTagName: string) => {
     if (hasAllowedEncodings && lcTagName === 'semantics') {
-      evt.allowedTags[lcTagName] = true;
+      return Optional.some(true);
+    } else if (lcTagName === 'annotation') {
+      return Optional.some(node.nodeType === NodeTypes.ELEMENT && hasValidEncoding(node));
+    } else if (Type.isArray(settings.allow_mathml_elements)) {
+      return Optional.some(settings.allow_mathml_elements.includes(lcTagName));
+    } else {
+      return Optional.none();
     }
+  };
+
+  purify.addHook('uponSanitizeElement', (node, evt) => {
+    // We know the node is an element as we have
+    // passed an element to the purify.sanitize function below
+    const lcTagName = evt.tagName ?? node.nodeName.toLowerCase();
+    const keepElementOpt = isValidElementOpt(lcTagName);
 
-    if (lcTagName === 'annotation') {
-      // We know the node is an element as we have
-      // passed an element to the purify.sanitize function below
-      const elm = node as Element;
-      const keepElement = hasValidEncoding(elm);
+    keepElementOpt.each((keepElement) => {
       evt.allowedTags[lcTagName] = keepElement;
-      if (!keepElement) {
-        elm.remove();
+      if (!keepElement && settings.sanitize) {
+        if (NodeType.isElement(node)) {
+          node.remove();
+        }
+      }
+    });
+  });
+
+  purify.addHook('uponSanitizeAttribute', (_node, event) => {
+    if (Type.isArray(settings.allow_mathml_attributes)) {
+      const keepAttribute = settings.allow_mathml_attributes.includes(event.attrName);
+
+      if (keepAttribute) {
+        event.forceKeepAttr = true;
```

Comment:
This doesn't seem right to me. We still want to check the node is a mathml element and also rather than `event.forceKeepAttr = true;` we should do the reverse and remove the non-allowed attribute instead. That seems safer to me at least.

Reply:
The node should already be checked by the node checking logic, We wouldn't need to check it a second time.

And if I remember right the default behavior is to remove many of the attributes, so not setting forcekeep just got them removed anyway, which wasn't very helpful.